lib/integrations/externalcloudaudit/long_term_storage.go

```diff
@@ -0,0 +1,161 @@
+	createLTSBucketClient
+}
+
+// EnsureLTSBucket ensures the long term storage bucket containing sessions and audit events is present and configured properly
+// This function checks if a bucket exists and is configured properly and creates one if it doesn't exist
+// If a bucket exists and doesn't have recommended settings configured we currently warn the user and return
+func EnsureLTSBucket(ctx context.Context, s3Client EnsureLTSBucketClient, req EnsureLTSBucketRequest) error {
+	// Check for existing bucket and it's existing configuration
+	// Don't return early if the error is bucket not found as that means we want to create it
+	err := checkLTSBucket(ctx, s3Client, req.BucketName, req.Region)
+	if err == nil || !isNoSuchBucketError(err) {
+		return trace.Wrap(err)
+	}
+
+	// Create LTS Bucket
+	err = createLTSBucket(ctx, s3Client, req.BucketName, req.Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+type checkLTSBucketClient interface {
+	// Gets the Object Lock configuration for a bucket.
+	GetObjectLockConfiguration(ctx context.Context, params *s3.GetObjectLockConfigurationInput, optFns ...func(*s3.Options)) (*s3.GetObjectLockConfigurationOutput, error)
+	// Retrieves OwnershipControls for an Amazon S3 bucket.
+	GetBucketOwnershipControls(ctx context.Context, params *s3.GetBucketOwnershipControlsInput, optFns ...func(*s3.Options)) (*s3.GetBucketOwnershipControlsOutput, error)
+	// Retrieves the PublicAccessBlock configuration for an Amazon S3 bucket.
+	GetPublicAccessBlock(ctx context.Context, params *s3.GetPublicAccessBlockInput, optFns ...func(*s3.Options)) (*s3.GetPublicAccessBlockOutput, error)
+	// Returns the Region the bucket resides in.
+	GetBucketLocation(ctx context.Context, params *s3.GetBucketLocationInput, optFns ...func(*s3.Options)) (*s3.GetBucketLocationOutput, error)
+}
+
+// checkLTSBucket checks for an existing bucket comparing its configuration to what's recommended
+func checkLTSBucket(ctx context.Context, s3Client checkLTSBucketClient, bucketName string, region string) error {
+	// check the region the bucket exists in
+	locationOutput, err := s3Client.GetBucketLocation(ctx, &s3.GetBucketLocationInput{Bucket: &bucketName})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if locationOutput.LocationConstraint != types.BucketLocationConstraint(region) {
+		return trace.Errorf("%s is not configured for the %s region", bucketName, region)
+	}
+
+	// Versioning is required by object lock so this covers checking if versioning is enabled or not
+	objectOutput, err := s3Client.GetObjectLockConfiguration(ctx, &s3.GetObjectLockConfigurationInput{Bucket: &bucketName})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if objectOutput.ObjectLockConfiguration.ObjectLockEnabled != types.ObjectLockEnabledEnabled {
+		// Is this an appropriate error code?
+		return trace.CompareFailed("Existing bucket does not have object lock enabled. Please specify a new bucket or enable object lock on existing one.")
```

Comment:
Go with BadParameter instead of CompareFailed, the latter is typically for failed CompareAndSwaps